### PR TITLE
Fix bug on disconnecting PayPal

### PR DESCRIPTION
**Pull request by rohitpaulk**
#4168

### Todo

- [x] fix accidentally failing tests
- [x] ~~lightly refactor `test_routes.py`~~ proved difficult due to fixture entanglement w/ `BillingHarness`(?)
- [x] fix (rename?) `test_delete_paypal_with_exchanges`
- [x] ensure we can readd a PayPal
- [x] ensure we can revive an old PayPal

### CODE REVIEW

**Comment by rohitpaulk** [comment]
Copying from #4168: 

Aha! 

screen shot 2016-10-31 at 6 37 24 pm

The `DELETE` statement was added by me in https://github.com/gratipay/gratipay.com/commit/0500105b9dc148855b7957963b3949d89a7cc3b4, and @whit537 added the comment above that in https://github.com/gratipay/gratipay.com/commit/e27a2283211bf9ade2121dec222714110e37b93b. 

I think the best way out here is to update the error to 'invalidated' irrespective of the `network` - and when user tries to add the route again for paypal, we revive the previous `exchange_route`.

**Comment by rohitpaulk** [comment]
I'm leaning towards "Update the ID to the next auto-incremented ID when reviving the previous exchange_route".

**Comment by rohitpaulk** [comment]
Nah, now "Change the condition for current_exchange_routes to not include routes that have error = 'invalidated'" makes more sense.

**Comment by rohitpaulk** [comment]
Dealing with the `invalidated` constant is turning out to be a pain - I think it is time to treat soft-deleted exchange routes as first-citizens by introducing a `is_deleted` field.

**Comment by rohitpaulk** [comment]
> Dealing with the invalidated constant is turning out to be a pain

This leads to bugs and special handling in many places.

- When a card is removed, we currently tell the user 'Your credit card is failing' (we should say that they haven't linked a credit card)

screen shot 2016-11-14 at 8 35 17 pm

We also end up showing them a notification with the same text.

- In the new functionality I'm going to introduce, we'll have to do a lot of logic in the application layer that could've been handled at the DB layer if we had a boolean `is_deleted` field.

**Comment by whit537** [comment]
Sentry: https://sentry.io/gratipay/gratipay-com/issues/189911878/.

**Comment by whit537** [comment]
@rohitpaulk Any chance you could move this forward, or perhaps find a way to fix #4168 as a manual one-off? 🙇

**Comment by whit537** [comment]
Rebased, was 1c6fa2829010ca9965779240c0ebbf2ec422c8da.

**Comment by whit537** [comment]
> revive the previous `exchange_route`

I agree.

> we're just picking the latest ID in that case

Yeah that part is weird. I'd expect a ctime/mtime in there to sort on.

> I think it is time to treat soft-deleted exchange routes as first-citizens by introducing a `is_deleted` field.

👍

**Comment by whit537** [comment]
Looks like we're pretty far along here. I'm picking this up ...

**Comment by whit537** [comment]
Rebased, was f8c90be61d78db4b002d0298d23ef80acb69fde6.

**Comment by whit537** [comment]
Started a todo in the description.

**Comment by whit537** [comment]
I think that's it! Ready for review @nobodxbodon @mattbk @JessaWitzel @rohitpaulk et al.

**Comment by mattbk** [comment]
Works for me after remembering to `make schema`.

File: gratipay/models/exchange_route.py [hunk 1]
```diff
@@ -69,20 +67,35 @@ def insert(cls, participant, network, address, error='', fee_cap=None, cursor=No
...
             braintree.PaymentMethod.delete(self.address)
-
-        # For Paypal, we remove the record entirely to prevent
-        # an integrity error if the user tries to add the route again
-        if self.network == 'paypal':
-            # XXX This doesn't sound right. Doesn't this corrupt history pages?
-            self.db.run("DELETE FROM exchange_routes WHERE id=%s", (self.id,))
-        else:
-            self.update_error('invalidated')
+        with self.db.get_cursor() as cursor:
```
Comment (nobodxbodon) [inline gratipay/models/exchange_route.py]: To confirm, we don't need to check if network is paypal anymore here?

File: gratipay/models/exchange_route.py [hunk 2]
```diff
@@ -69,20 +67,35 @@ def insert(cls, participant, network, address, error='', fee_cap=None, cursor=No
...
+        self.set_attributes(is_deleted=False)
 
     def update_error(self, new_error):
         id = self.id
         old_error = self.error
-        if old_error == 'invalidated':
+
+        if self.is_deleted:
             return
+
```
Comment (nobodxbodon) [inline gratipay/models/exchange_route.py]: nit: move this check to the top of method?

File: gratipay/models/exchange_route.py [hunk 3]
```diff
@@ -53,12 +49,14 @@ def from_address(cls, participant, network, address):
         return r
 
     @classmethod
-    def insert(cls, participant, network, address, error='', fee_cap=None, cursor=None):
+    def insert(cls, participant, network, address, error='', fee_cap=None, cursor=None,
+                                                                                 is_deleted=False):
```
Comment (nobodxbodon) [inline gratipay/models/exchange_route.py]: as we don't really "delete' the route, shall we name it (and in the database) something like 'active'?

File: tests/py/test_routes.py [hunk 4]
```diff
@@ -26,34 +24,77 @@ def test_associate_and_delete_valid_card(self):
...
+
+        assert not self.roman.has_payout_route
+
+    def test_add_new_paypal_address(self):
+        self.add_and_verify_email(self.roman, 'roman@gmail.com')
+        self.add_and_verify_email(self.roman, 'candle@gmail.com')
+
+        self.hit('roman', 'associate', 'paypal', 'roman@gmail.com')
+        self.hit('roman', 'delete', 'paypal', 'roman@gmail.com')
+        self.hit('roman', 'associate', 'paypal', 'candle@gmail.com')
```
Comment (nobodxbodon) [inline tests/py/test_routes.py]: could you extract above 5 lines to separate method, and reuse in test case below?

File: www/~/%username/routes/associate.json.spt [hunk 5]
```diff
@@ -40,7 +40,13 @@ if change:
     elif network == 'paypal':
         if address not in participant.get_verified_email_addresses():
             raise Response(400, _("Only verified email addresses allowed."))
-        ExchangeRoute.insert(participant, network, address)
+        try:
+            ExchangeRoute.insert(participant, network, address)
+        except IntegrityError as e:
+            if e.pgcode != errorcodes.UNIQUE_VIOLATION:
+                raise
+            existing = ExchangeRoute.from_address(participant, network, address)
```
Comment (nobodxbodon) [inline www/~/%username/routes/associate.json.spt]: if I'm not mistaken, this occurs when insertion fails. Could we check before insertion instead? And one nit: rename existing to existing_route?

File: gratipay/models/exchange_route.py [hunk 6]
```diff
@@ -69,20 +67,35 @@ def insert(cls, participant, network, address, error='', fee_cap=None, cursor=No
...
-        if self.network == 'paypal':
-            # XXX This doesn't sound right. Doesn't this corrupt history pages?
-            self.db.run("DELETE FROM exchange_routes WHERE id=%s", (self.id,))
-        else:
-            self.update_error('invalidated')
+        with self.db.get_cursor() as cursor:
+            self.db.run("UPDATE exchange_routes SET is_deleted=true WHERE id=%s", (self.id,))
+            payload = dict( id=self.participant.id
+                          , exchange_route=self.id
+                          , action='invalidate route'
```
Comment (nobodxbodon) [inline gratipay/models/exchange_route.py]: not quite familiar with this 'action' field. I imagine there are a set of actions, each of which should be constants. IMO they are better to be defined in one place, but feel free to ignore if it involves too much refactoring.

**Comment by rohitpaulk** [comment]
Rebased on master.

File: gratipay/models/exchange_route.py (hunk 3 above)
Comment (rohitpaulk) [inline gratipay/models/exchange_route.py]: Hmm, I don't think it is worth adding `is_deleted` to the `insert` function. No one (other than in tests) is going to insert a deleted route, are they

File: www/~/%username/routes/associate.json.spt (hunk 5 above)
Comment (rohitpaulk) [inline www/~/%username/routes/associate.json.spt]: Checking before insertion makes us prone to race conditions (two people check, both find no route, both try to write, one fails). 

I'm changing `existing` to `existing_route`

File: gratipay/models/exchange_route.py (hunk 3 above)
Comment (rohitpaulk) [inline gratipay/models/exchange_route.py]: In fact, we should only be accepting `participant`, `network`, `address` and `fee_cap` when creating an exchange route. Routes are never created with errors, and routes are never created with `is_deleted=true`

File: gratipay/models/exchange_route.py (hunk 1 above)
Comment (rohitpaulk) [inline gratipay/models/exchange_route.py]: No, we don't. We've now changed every route to be soft-deleted

File: gratipay/models/exchange_route.py (hunk 6 above)
Comment (rohitpaulk) [inline gratipay/models/exchange_route.py]: Event properties are logged freeform JSON, I don't think we reuse these values between multiple functions anywhere

File: tests/py/test_routes.py (hunk 4 above)
Comment (rohitpaulk) [inline tests/py/test_routes.py]: IMO that will be more confusing to read

**Comment by rohitpaulk** [comment]
Ready for review, @nobodxbodon

**Pull request merged**